### PR TITLE
Refs #32819 -- Added id to ErrorList class and template.

### PR DESCRIPTION
**Pull request by smithdc1**
#### Trac ticket number

ticket-32819

#### Branch description
Step 1 of ticket 32819 is to add an id to the error list template so the `aria-describedby` has something to point to. See #17520

#### Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/forms/forms.py [hunk 1]
```diff
@@ -252,7 +252,7 @@ def non_field_errors(self):
             self.error_class(error_class="nonfield", renderer=self.renderer),
         )
 
-    def add_error(self, field, error):
+    def add_error(self, field, error, field_id=None):
```
Comment (sarahboyce) [inline django/forms/forms.py:255]: I feel like we could do something like:
```diff
--- a/django/forms/forms.py
+++ b/django/forms/forms.py
@@ -252,7 +252,7 @@ class BaseForm(RenderableFormMixin):
             self.error_class(error_class="nonfield", renderer=self.renderer),
         )
 
-    def add_error(self, field, error, field_id=None):
+    def add_error(self, field, error):
         """
         Update the content of `self._errors`.
 
@@ -298,6 +298,7 @@ class BaseForm(RenderableFormMixin):
                         error_class="nonfield", renderer=self.renderer
                     )
                 else:
+                    field_id = self.__getitem__(field).auto_id
                     self._errors[field] = self.error_class(
                         renderer=self.renderer,
                         field_id=field_id,
@@ -338,7 +339,7 @@ class BaseForm(RenderableFormMixin):
                     value = getattr(self, "clean_%s" % name)()
                     self.cleaned_data[name] = value
             except ValidationError as e:
-                self.add_error(name, e, bf.auto_id)
+                self.add_error(name, e)
 
     def _clean_form(self):
```

If we don't do this, we need doc updates to `.. method:: Form.add_error` and anyone who has used `.add_error` will be missing the id I suppose

File: docs/ref/forms/api.txt [hunk 2]
```diff
@@ -1098,6 +1105,10 @@ Customizing the error list format
     overriding the default template, see also
     :ref:`overriding-built-in-form-templates`.
 
+    .. versionadded:: 5.2
+
+        The ``field_id`` attributed was added.
```
Comment (sarahboyce) [inline docs/ref/forms/api.txt:1110]: I think I prefer something like this for the versionadd/changed:

```diff
     methods.
 
+    .. versionchanged:: 5.2
+
+        The ``field_id`` argument was added.
+
     .. attribute:: error_class
 
         The CSS classes to be used when rendering the error list. Any provided
@@ -1045,6 +1049,8 @@ Customizing the error list format
 
     .. attribute:: field_id
 
+        .. versionadded:: 5.2
+
         An ``id`` for the field for which the errors relate. This allows an
```

File: docs/ref/forms/api.txt [hunk 3]
```diff
@@ -1043,6 +1043,13 @@ Customizing the error list format
         Defaults to ``None`` which means to use the default renderer
         specified by the :setting:`FORM_RENDERER` setting.
 
+    .. attribute:: field_id
```
Comment (sarahboyce) [inline docs/ref/forms/api.txt:1046]: Do you think it's worth mentioning that this defaults to the auto_id for field errors and no id is associated to non-field errors?

**Review by sarahboyce — COMMENTED** [review]
Thank you for pulling this out @smithdc1 :star:

File: docs/ref/forms/api.txt (hunk 3 above)
Comment (smithdc1) [reply docs/ref/forms/api.txt:1050]: I made some changes here, could you have another look? I'm not sure if we need to go further and mention non-field errors.

Comment (sarahboyce) [reply docs/ref/forms/api.txt:1050]: This looks great thank you!

**Review by sarahboyce — COMMENTED** [review]
Checked the docs and the rendering of errors seems to always have `auto_id=False` so I don't think we need any more doc updates :confetti_ball: 

I have a suggestion on the release note but otherwise I think this is good to go :+1:

File: docs/releases/5.2.txt [hunk 4]
```diff
@@ -230,6 +230,9 @@ Forms
 * The new :class:`~django.forms.TelInput` form widget is for entering telephone
   numbers and renders as ``<input type="tel" ...>``.
 
+* :class:`~django.forms.ErrorList` now accepts a ``field_id`` argument and an
+  ``id`` HTML element is added to the default ``ErrorList`` template.
```
[the remark this answers is not in the document]
Comment (smithdc1) [reply docs/releases/5.2.txt:234]: I like your second suggestion.

File: docs/ref/forms/api.txt [hunk 5]
```diff
@@ -1416,7 +1430,7 @@ Methods of ``BoundField``
 
 .. method:: BoundField.render(template_name=None, context=None, renderer=None)
 
-    The render method is called by ``as_field_group``. All arguments are 
+    The render method is called by ``as_field_group``. All arguments are
```
[the remark this answers is not in the document]
Comment (smithdc1) [reply docs/ref/forms/api.txt:1433]: I'll rebase 👍

**Pull request merged**